### PR TITLE
AST: throw expected error messages

### PR DESCRIPTION
**Pull request by helixbass**
@GeoffreyBooth PR to throw (almost all) expected error messages during AST generation (in addition to during compile-to-JS)

### CODE REVIEW

File: src/nodes.coffee [hunk 1]
```diff
@@ -3387,11 +3437,11 @@ exports.Assign = class Assign extends Base
   unfoldSoak: (o) ->
     unfoldSoak o, this, 'variable'
 
-  addScopeVariables: (o, {checkAssignability = yes} = {}) ->
+  addScopeVariables: (o, {allowAssignmentToExpansion = no} = {}) ->
```
Comment (helixbass) [inline src/nodes.coffee]: Had to refine the disabling of `checkAssignability` during AST generation to only specifically disable checking `Expansion`'s since it _should_ check assignability for things like `null` (eg `[null] = b`)

File: test/error_messages.coffee [hunk 2]
```diff
@@ -8,10 +8,10 @@ errCallback = (expectedErrorFormat) -> (err) ->
   err.colorful = no
   eq expectedErrorFormat, "#{err}"
   yes
-assertErrorFormat = (code, expectedErrorFormat) ->
+assertErrorFormatNoAst = (code, expectedErrorFormat) ->
```
Comment (helixbass) [inline test/error_messages.coffee]: Adjusted so the "default" `assertErrorFormat()` includes AST checking, and renamed the existing `assertErrorFormat()` to `assertErrorFormatNoAst()` (for cases that don't also throw an error during AST generation)

File: test/error_messages.coffee [hunk 3]
```diff
@@ -1122,15 +1122,15 @@ test "cannot export * without a module to export from", ->
   '''
 
 test "imports and exports must be top-level", ->
```
Comment (helixbass) [inline test/error_messages.coffee]: This is the only one (at least from `test/error_messages.coffee`) that still isn't being thrown during AST generation (but in theory could) - the logic for throwing an error on "non-top-level" imports/exports is based on `o.indent`, which doesn't get set during AST generation

Seems like there could be a different mechanism not based on indent level to detect when you're not at the top level in general?

Comment (GeoffreyBooth) [reply test/error_messages.coffee]: Does `if o.level > LEVEL_TOP` not work?

File: src/nodes.coffee [hunk 4]
```diff
@@ -3175,6 +3201,10 @@ exports.ImportDeclaration = class ImportDeclaration extends ModuleDeclaration
     code.push @makeCode ';'
     code
 
+  astNode: (o) ->
+    o.importedSymbols = []
```
Comment (GeoffreyBooth) [inline src/nodes.coffee]: Why is this here?

Comment (helixbass) [reply src/nodes.coffee]: This `o.importedSymbols` gets used to track imported names eg for throwing an error if you try and import the same name more than once (eg `import {foo, foo} from 'bar'`) in `ImportSpecifier::addIdentifierToScope()`

File: test/error_messages.coffee (hunk 3 above)
Comment (helixbass) [reply test/error_messages.coffee]: I think that wouldn't work since `level` really only reflects the "compilation level" *within a statement*

ie block-level statements always get compiled at `LEVEL_TOP`, eg here the import statement gets compiled at `LEVEL_TOP`:
```
if foo
  import { bar } from 'lib'
```

But it does seem like there could be a separate concept of "program top-level" (meaning I guess "a statement in the root `Block`"?)

It seems like that could get set on `o`. If we ended up tracking block scopes eg to be able to generate `const`/`let` where possible per your other comment, it seems like that would suffice

Comment (GeoffreyBooth) [reply test/error_messages.coffee]: > It seems like that could get set on `o`. If we ended up tracking block scopes eg to be able to generate `const`/`let` where possible per your other comment, it seems like that would suffice

Is tracking block scopes something that’s easy to add? If so, absolutely let’s do that, I’d love to start outputting `let`/`const` 😄 

Does a node know its parents? Obviously it knows its children, but if it also can easily find its parents then you can see if there are any `Block`s among its parents.

This is also not the most important check. The runtime will throw if anyone tries to put an `import` in not-top-level code in executable JavaScript (as opposed to JS further transpiled by some other build tool). We could just stop checking for this in CoffeeScript entirely, or check only on JS output with a comment saying why it’s hard to do this check on AST output, and leave it at that.

File: src/nodes.coffee (hunk 4 above)
Comment (GeoffreyBooth) [reply src/nodes.coffee]: I see, you’re duplicating what’s in `compileNode`. I guess we can’t just put this in the constructor to avoid the duplication?

I also don’t care all that much, it’s one line.

Comment (helixbass) [reply src/nodes.coffee]: Ya in general I took the approach of trying to DRY things up by sharing small helper methods between the `compileNode()` path and the AST generation path, but here this didn't seem worth trying to encapsulate

File: test/error_messages.coffee (hunk 3 above)
Comment (helixbass) [reply test/error_messages.coffee]: > Is tracking block scopes something that’s easy to add? If so, absolutely let’s do that, I’d love to start outputting `let`/`const` 😄

I'm not sure, but I'd be up for doing a deeper dive on how scopes work sometime soon

> Does a node know its parents?

Ya I was thinking something similar. Currently I don't think there's any way to walk "up" the node class object tree. And I can't picture a simple way to add "parent links" to enable that

> This is also not the most important check

Agreed, I think it's fine for this to remain something that only gets flagged during compile-to-JS for the time being

> or check only on JS output with a comment saying why it’s hard to do this check on AST output

Sure, added a comment to `ModuleDeclaration::checkScope()`
